cpp/include/rapidsmpf/streaming/core/message.hpp

```diff
@@ -0,0 +1,217 @@
+
+    /**
+     * @brief Resets the message to an empty state.
+     *
+     * If this is the last shared owner, the stored payload is deallocated.
+     * Otherwise, the payload remains untouched and shared with other owners.
+     *
+     * @note After this call, the message becomes empty.
+     */
+    void reset() noexcept {
+        return payload_.reset();
+    }
+
+    /**
+     * @brief Whether the message currently holds a payload.
+     *
+     * @return `true` if the message is empty; otherwise, `false`.
+     */
+    [[nodiscard]] bool empty() const noexcept {
+        if (payload_) {
+            std::lock_guard<std::mutex> lock(payload_->mutex);
+            return !payload_->data.has_value();
+        }
+        return true;
+    }
+
+    /**
+     * @brief Check whether the stored payload has type `T`.
+     *
+     * @tparam T Expected payload type.
+     * @return `true` if the payload type matches; otherwise, `false`.
+     * @throws std::invalid_argument if the message is empty.
+     */
+    template <typename T>
+    [[nodiscard]] bool holds() const {
+        auto lock = lock_payload();
+        return payload_->data.type() == typeid(std::shared_ptr<T>);
+    }
+
+    /**
+     * @brief Access the payload by const reference.
+     *
+     * The returned reference remains valid until the message is reset or the payload is
+     * released.
+     *
+     * @tparam T Payload type.
+     * @return Const reference to the stored payload.
+     * @throws std::invalid_argument if the message is empty or the type mismatches.
+     */
+    template <typename T>
+    T const& get() const {
+        auto [ret, lock] = get_ptr_and_lock<T>();
+        return *ret;
+    }
+
+    /**
+     * @brief Move the payload out of the message and empty it.
+     *
+     * Transfers ownership of the payload out of the message. This operation is
+     * only allowed if the message is the sole owner of the payload.
+     *
+     * @tparam T Payload type to extract.
+     * @return The payload moved out of the message.
+     *
+     * @throws std::invalid_argument if:
+     *         - The message is empty.
+     *         - The stored type does not match `T`.
+     *         - There are multiple shared references to the same payload.
+     *
+     * @note After this call, the message is empty.
+     */
+    template <typename T>
+    T release() {
+        // If this is the last reference, `reset()` deallocates `payload_` thus
+        // we have to move the payload to a new shared_ptr before resetting.
+        auto ret = [&]() -> std::shared_ptr<T> {
+            auto [ptr, lock] = get_ptr_and_lock<T>();
+            RAPIDSMPF_EXPECTS(
+                payload_.use_count() == 1,
+                "release() requires this to be the sole owner of the payload",
+                std::invalid_argument
+            );
+            return std::move(ptr);
+        }();
+        reset();
+        return std::move(*ret);
```

Comment:
mutex is not locked at this point. So, someone can make a shallow_copy before this `reset`